File: apps/mybookkeeper/frontend/src/shared/types/insurance/insurance-policy-summary.ts
import type { InsurancePremiumFrequency } from "@/shared/types/insurance/insurance-premium-frequency";

/**
 * Summary row for the insurance policy list.
 *
 * Mirrors ``schemas/insurance/insurance_policy_summary.py``.
 *
 * ``wind_hail_deductible_pct`` arrives as a JSON *string* because the backend
 * column is ``Numeric`` — see ``shared/lib/insurance-policy-format.ts``, which
 * is the only place it is parsed.
 */
export interface InsurancePolicySummary {
  id: string;
  property_id: string;
  property_name: string | null;
  policy_name: string;
  carrier: string | null;
  effective_date: string | null;
  expiration_date: string | null;
  coverage_amount_cents: number | null;
  premium_cents: number | null;
  premium_frequency: InsurancePremiumFrequency | null;
  deductible_cents: number | null;
  wind_hail_deductible_pct: string | null;
  /** Derived server-side from premium + frequency; never sent on write. */
  annual_premium_cents: number | null;
  created_at: string;
  updated_at: string;
}

File: apps/mybookkeeper/backend/app/schemas/insurance/insurance_policy_summary.py
"""Summary row for the insurance policy list view."""
from __future__ import annotations

import datetime as _dt
import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, computed_field

from app.services.insurance.premium_math import annual_premium_cents, annual_total_cents


class InsurancePolicySummary(BaseModel):
    id: uuid.UUID
    property_id: uuid.UUID
    # Resolved by the service so the list can name the building without the
    # client holding its own copy of the property table.
    property_name: str | None = None
    policy_name: str
    carrier: str | None = None
    effective_date: _dt.date | None = None
    expiration_date: _dt.date | None = None
    coverage_amount_cents: int | None = None
    premium_cents: int | None = None
    premium_frequency: str | None = None
    fees_and_taxes_cents: int | None = None
    deductible_cents: int | None = None
    wind_hail_deductible_pct: Decimal | None = None
    created_at: _dt.datetime
    updated_at: _dt.datetime

    @computed_field  # type: ignore[prop-decorator]
    @property
    def annual_premium_cents(self) -> int | None:
        """The premium restated as a yearly total.

        Derived rather than stored so it can never disagree with the amount and
        period it comes from, and computed here rather than in the two callers
        so the list and the detail view cannot annualise differently.

        The premium ALONE — fees and taxes are excluded deliberately. This is
        the figure the overpaying comparison measures against a county average,
        and county averages are premiums; adding the paperwork to one side of
        that comparison would report a fairly-priced policy as 15% over market.
        """
        return annual_premium_cents(self.premium_cents, self.premium_frequency)

    @computed_field  # type: ignore[prop-decorator]
    @property
    def annual_total_cents(self) -> int | None:
        """A year of premium plus the term's fees and taxes.

        What the policy actually costs to hold, as opposed to what the coverage
        is priced at. Derived here for the same reason as the annual premium:
        two callers deriving it separately is two chances to disagree.
        """
        return annual_total_cents(
            self.premium_cents, self.premium_frequency, self.fees_and_taxes_cents,
        )

    model_config = ConfigDict(from_attributes=True)
